Content refs

emirotin opened the pull request:
I ave just finished testing it so it may look dirty somewhere but overall it looks like it's working!

Here are the highlights:

* to define the reference to another category item in the JSON schema you use `{ type: 'string', $subtype: 'ref', $category: 'trivia' }`
* the actual ref is stored as `"##ref(trivia-question-9l9hH~)"`
* the content picker UI is the same from the flows dialog
* I made couple of important fixes (some to my own bugs, and one to allow non-object values, like form of single string, or an array)
* the forms extra methods (like `computeFormData` now get the 2nd arg, the function of the same name, that can be used to serialize other form elements, like in here: `return Promise.map(formData, computeFormData.bind(null, 'trivia'))`)
* I've changed the source maps generation because the previous one didn't work for me at all (I wasn't able to properly step in Chrome). The new options are from the webpack docs recommendations and also seem to be a little bit faster

emirotin committed:
extract content picker widget into a reusable component

emirotin committed:
move generic content components out of views

emirotin committed:
fix array content creation bug

emirotin committed:
content picker integrated into the form, buggy (resets to undefined, wtf)

emirotin committed:
a bunhc of fixes for source maps and content creation

emirotin committed:
extra sample bot data

emirotin committed:
fix content picker search count

emirotin committed:
fix something, lol

emirotin committed:
fix unneeded fetches

emirotin committed:
fixes here and there, everything is complicated :)

emirotin committed:
working refs saving, yay!

emirotin committed:
more sample data

epaminond commented on src/content/service.js:
```diff
@@ -130,18 +136,70 @@ module.exports = async ({ botfile, projectLocation, logger, ghostManager }) => {
+
   const fillComputedProps = async (category, formData) => {
-    if (_.isNil(formData) || !_.isObject(formData)) {
+    if (formData == null) {
```
@emirotin , could we use strict comparisons (`===`) or keep `_.isNil`?

emirotin replied:
Oh, it's  a matter of style of course. It's intentionally non-strict everywhere when I compare to `null` (and mean to also compare to undefined,  because it's the native equivalent of `_.isNil`. Basically I dislike using utility methods where native equivalents exist and are short and concise. But can go back to `_.isNil` if you prefer

epaminond replied:
What about shorter `!formData`? )

emirotin replied:
That will wrongly catch empty strings and numeric zeroes which are valid values now

slvnperron commented:
Amazing!! That works well!

I'm only facing one issue: when you refresh the page and load an item with nested content item previously saved, the nested items doesn't seem to render:

![image](https://user-images.githubusercontent.com/1315508/35310362-045a1e6a-007f-11e8-9382-b574833da2ab.png)

slvnperron commented:
An other issue with multiple overlays overriding each others?

![pr_385](https://user-images.githubusercontent.com/1315508/35310475-9c312b16-007f-11e8-8568-3ef214c7dd64.gif)

slvnperron commented on dev-bot/forms_data/trivia-collection.json:
```diff
@@ -0,0 +1,14 @@
+  },
+  {
+    "id": "trivia-collection-AxRaw~",
+    "formData": ["##ref(trivia-question-8CXNiN)"],
```
Love the syntax

slvnperron commented on src/content/service.js:
```diff
@@ -130,18 +136,70 @@ module.exports = async ({ botfile, projectLocation, logger, ghostManager }) => {
       }
     })
 
+  const resolveRefs = Promise.method(data => {
```
Could be made `async` instead?

slvnperron commented on src/web/components/Content/Select/index.js:
```diff
@@ -31,32 +33,33 @@ const initialData = {
-        this.setState({ show: true, activeItemIndex: 0 })
-        setImmediate(() => moveCursorToEnd(this.searchInput))
-      })
+    window.botpress.pickContent = ({ categoryId = null } = {}, callback) => {
```
We should probably define this in a higher level component to allow multiple simultaneous pickers

slvnperron reviewed (approved):
The code looks all good 👍

emirotin committed:
fix GCM docs

emirotin commented:
Hm, the 2nd issue is understandable, it's the singleton Select biting us indeed.

The 1st one is unexpected though, but surely I'll find a way to fix it.

emirotin committed:
no more singleton content selector

emirotin committed:
fix RefWidget rendering with existing content

emirotin commented:
Fixed both issues. The 2nd one was because I didn't account for `componentWillReceiveProps` not called for the initial mount with props, and thus the content item details were never fetched.

For the 1st one I've implemented the proper modals queue instead of the singleton. But I want to debug it a little more and also remove the calls to pickContent from our code to keep it more react-ish

emirotin commented:
Will address the code comments today and also update the docs.

emirotin committed:
throw if referenced content not found

emirotin commented:
@slvnperron can you give it some more testing please?

slvnperron commented on webpack.js:
```diff
@@ -7,7 +7,7 @@ const UglifyJSPlugin = require('uglifyjs-webpack-plugin')
 
 const nodeConfig = {
-  devtool: process.env.NODE_ENV === 'production' ? 'source-map' : 'eval-cheap-module-source-map',
+  devtool: 'source-map',
```
This slowed my build time from 10-15s to 30-35s in dev. What's your build time like?

emirotin replied:
Cold prod compilation — 23s, dev — 33s.
My problem with the previous option was it gave me zero hints for the errors thrown from the generators (like transpiled async/await code).
Let's try to find the sweet spot, but before the change I basically wasn't able to trace the errors.

For the browser change I also wasn't able to step through the code in the debugger before, now it works like a charm.

slvnperron replied:
OK thanks, our build time are similar. I'll open a out-of-`x` issue about optimizing this. Let's keep it as is

slvnperron commented:
Both issues are fixed on my end and this is working quite well!!

I only found one more issue, which happens when I try to validate the form by pressing "Enter" (which I always do), which closes all Content windows instead of only the one with focus.

Video:

![pr_385_2](https://user-images.githubusercontent.com/1315508/35359324-a338d9f6-0127-11e8-804f-ad25b481f9fe.gif)

slvnperron commented:
@emirotin I opened an issue about this instead and merged this so we don't get the branch behind and I can start using this right away
